### DISCUSSION TITLE
LND stopped working after restart (stuck in a loop)

### DISCUSSION THREAD

**Discussion by guitarchair** [lightningnetwork/lnd#8935] [Troubleshooting] [answered]
I just restarted my node and lnd stopped working, the logs show me this:

2024-07-24 10:49:14.274 [INF] LTND: Version: 0.18.0-beta commit=v0.18.0-beta, build=production, logging=default, debuglevel=info
2024-07-24 10:49:14.274 [INF] LTND: Active chain: Bitcoin (network=mainnet)
2024-07-24 10:49:14.275 [INF] RPCS: RPC server listening on 0.0.0.0:10009
2024-07-24 10:49:14.283 [INF] RPCS: gRPC proxy started at 0.0.0.0:8080
2024-07-24 10:49:14.283 [INF] LTND: Opening the main database, this might take a few minutes...
2024-07-24 10:49:14.283 [INF] LTND: Opening bbolt database, sync_freelist=false, auto_compact=false
2024-07-24 10:50:06.026 [WRN] LTND: Config 'bitcoin.active' is deprecated, please remove it
2024-07-24 10:50:06.026 [INF] LTND: Version: 0.18.0-beta commit=v0.18.0-beta, build=production, logging=default, debuglevel=info
2024-07-24 10:50:06.026 [INF] LTND: Active chain: Bitcoin (network=mainnet)
2024-07-24 10:50:06.027 [INF] RPCS: RPC server listening on 0.0.0.0:10009
2024-07-24 10:50:06.032 [INF] RPCS: gRPC proxy started at 0.0.0.0:8080
2024-07-24 10:50:06.032 [INF] LTND: Opening the main database, this might take a few minutes...
2024-07-24 10:50:06.032 [INF] LTND: Opening bbolt database, sync_freelist=false, auto_compact=false

**Comment by guitarchair** [comment]
I now tried to copy the backup of the /data dir I have (basically deleted the node config I had), and now I can at least start the lnd app. Maybe it will solve the issue after I try to restart again (I'll try that, after it's done rescanning).
I'll update this post later on with the results.

**Reply by guggero** [reply]
Did you edit the configuration file? Sounds like something went wrong there.
Feel free to post it here so I can take a look (editing out any passwords of course).

**Reply by guitarchair** [reply]
Yes I edited the file but noticed that it didn't change anything in that file after a restart (umbrel-lnd.conf). Maybe that's what broke the node? Strange, because it's as if no modification was done to the file, even after I saved it.

So now, I deleted the /data dir and copied it from the backup (which had the same single channel open, with no additional transactions made). And now, unfortunately, after restarting the node it shows it's closing this single channel I had.
Why is that? I thought a force-close is only done when there's a discrepancy in channel balance... I need to understand why so I won't repeat this mistake in the future.

The plan now is to try changing the lnd.conf file as described here:
https://community.umbrel.com/t/can-t-access-lnd-conf/13502/2
Which basically says I need to create the lnd.conf file and add the lines I want, and umbrel will append them to the general umbrel-lnd.conf file

After that, I'll reopen this channel..

**Reply by guggero** [reply]
https://github.com/lightningnetwork/lnd/blob/master/docs/safety.md#file-based-backups

**Reply by guitarchair** [reply]
Thank you. I was following the node migration guide:
https://docs.lightning.engineering/lightning-network-tools/lnd/migrating-lnd#start-lnd
You can see that it only warns about not running both nodes at the same time. I think it should also include the warning about not reusing the same files used for the migration (which worked the first time), because of the change in the channel.db file.
I don't know if you have access to modify that page, but it could have saved me from thinking "it'll be just like migrating a second time"

**Reply by guitarchair** [reply]
I uninstalled LND and started it from fresh completely (only reusing the seed). And after the installation it works and the node is up, but after I restart the node, lnd won't load, stuck in a loop again. Something about restarting lnd after it's installed makes it fail. Here is the log, goes in this endless loop. The bitcoind is fully synced.

2024-07-24 22:32:38.627 [WRN] LTND: Config 'bitcoin.active' is deprecated, please remove it
2024-07-24 22:32:38.627 [INF] LTND: Version: 0.18.0-beta commit=v0.18.0-beta, build=production, logging=default, debuglevel=info
2024-07-24 22:32:38.649 [INF] LTND: Active chain: Bitcoin (network=mainnet)
2024-07-24 22:32:38.650 [INF] RPCS: RPC server listening on 0.0.0.0:10009
2024-07-24 22:32:38.661 [INF] RPCS: gRPC proxy started at 0.0.0.0:8080
2024-07-24 22:32:38.662 [INF] LTND: Opening the main database, this might take a few minutes...
2024-07-24 22:32:38.716 [INF] LTND: Opening bbolt database, sync_freelist=false, auto_compact=false
2024-07-24 22:32:38.693 [ERR] RPCS: [/lnrpc.Lightning/GetInfo]: waiting to start, RPC services not available
2024-07-24 22:32:38.719 [INF] LTND: Shutdown complete

**Reply by guggero** [reply]
Is this log from the docker console or from the log file? I think there might be more in the docker output, as the actual error might only be printed to standard out and not into the log file...

**Reply by guitarchair** [reply]
It's from the log file, and the other lightning related docker logs don't show anything problematic, just started and waiting for the lnd to start... any other way to get more detailed logs?

umbrel@umbrel:~$ sudo docker ps | grep light
1b9c142e63b8   lightninglabs/lnd:v0.18.0-beta      "lnd --configfile=/d…"     13 seconds ago   Restarting (2) 2 seconds ago                                                                                                                                lightning_lnd_1
f3ec453afc2c   getumbrel/umbrel-lightning:v1.2.0   "docker-entrypoint.s…"     13 seconds ago   Up 11 seconds                  3006/tcp                                                                                                                     lightning_app_1
cc0e86a2f2c2   getumbrel/tor:0.4.7.8               "tor"                      13 seconds ago   Up 11 seconds                                                                                                                                               lightning_tor_1
d64e18447b95   getumbrel/app-proxy:1.0.0           "docker-entrypoint.s…"     13 seconds ago   Up 12 seconds                  0.0.0.0:2101->2101/tcp, :::2101->2101/tcp                                                                                    lightning_app_proxy_1

**Reply by guggero** [reply]
Can you do `docker logs lightning_lnd_1`?

**Reply by guitarchair** [reply]
That's the command I used for the logs I posted. It's the exact same logs as in "lnd.log"

**Reply by guggero** [reply]
okay, then I think something's wrong with your config file. can you post it please?

**Reply by guitarchair** [reply]
# This file is automatically generated by Umbrel, any changes will be overwritten.
# Generated at: 2024-07-24T22:55:27.423Z
tlsautorefresh=true
sync-freelist=false
payments-expiration-grace-period=30m
maxpendingchannels=1
minchansize=20000
maxchansize=16777215
coop-close-target-confs=6
stagger-initial-reconnect=false
max-cltv-expiry=2016
max-commit-fee-rate-anchors=10
accept-keysend=false
accept-amp=true
gc-canceled-invoices-on-startup=false
gc-canceled-invoices-on-the-fly=false
allow-circular-route=false
color=#3399FF
bitcoin.defaultchanconfs=2
bitcoin.basefee=1000
bitcoin.feerate=1
bitcoin.timelockdelta=80
tor.skip-proxy-for-clearnet-targets=true
tor.streamisolation=false
watchtower.active=false
wtclient.active=false
wtclient.sweep-fee-rate=10
routerrpc.minrtprob=0.01
routerrpc.apriori.hopprob=0.6
routerrpc.apriori.weight=0.5
routerrpc.apriori.penaltyhalflife=1h
routerrpc.attemptcost=100
routerrpc.attemptcostppm=1000
routerrpc.maxmchistory=1000
caches.channel-cache-size=20000
protocol.wumbo-channels=false
db.bolt.auto-compact=false
db.bolt.auto-compact-min-age=168h
rpcmiddleware.enable=true
routing.strictgraphpruning=false

**Reply by guggero** [reply]
Okay, that looks good. Did you add any custom config through the Umbrel UI?
Can you perhaps check if there's something more in the actual log file?
I think it should be somewhere like this: `${EXPORTS_APP_DIR}/data/lnd/logs` (not sure what `EXPORTS_APP_DIR` is set to, but you should be able to figure that out with `env | grep EXPORTS`.

**Reply by guitarchair** [reply]
I didn't touch anything. Just installed LND, saw that it was online up and running, then restarted it...
And there is no EXPORTS in the env output, so not sure what you meant, but the full path for the log is:
/data/umbrel-os/home/umbrel/umbrel/app-data/lightning/data/lnd/logs/bitcoin/mainnet/lnd.log
And under the "logs" dir there's only the "bitcoin" dir. No hidden dirs either.
The lnd.log file has a clean shutdown before this loop that gets stuck:


2024-07-24 22:28:13.805 [INF] CNCT: ChainArbitrator shutting down...
2024-07-24 22:28:13.805 [INF] NTFN: Cancelling epoch notification, epoch_id=8
2024-07-24 22:28:13.818 [INF] FNDG: Funding manager shutting down...
2024-07-24 22:28:13.890 [INF] BRAR: Breach arbiter shutting down...
2024-07-24 22:28:13.901 [INF] UTXN: UTXO nursery shutting down...
2024-07-24 22:28:13.902 [INF] NTFN: Cancelling epoch notification, epoch_id=5
2024-07-24 22:28:13.902 [INF] DISC: Authenticated gossiper shutting down...
2024-07-24 22:28:13.902 [INF] DISC: Authenticated Gossiper is stopping
2024-07-24 22:28:13.902 [INF] NTFN: Cancelling epoch notification, epoch_id=9
2024-07-24 22:28:14.002 [INF] DISC: Authenticated Gossiper stopped
2024-07-24 22:28:14.031 [INF] SWPR: Sweeper shutting down...
2024-07-24 22:28:14.031 [INF] NTFN: Cancelling epoch notification, epoch_id=4
2024-07-24 22:28:14.031 [INF] SWPR: TxPublisher stopping...
2024-07-24 22:28:14.032 [INF] NTFN: Cancelling epoch notification, epoch_id=3
2024-07-24 22:28:14.032 [INF] CHNF: ChannelNotifier shutting down...
2024-07-24 22:28:14.032 [INF] PRNF: PeerNotifier shutting down...
2024-07-24 22:28:14.032 [INF] HSWC: HtlcNotifier shutting down...
2024-07-24 22:28:14.032 [INF] CHBU: chanbackup.SubSwapper shutting down...
2024-07-24 22:28:14.032 [INF] NTFN: bitcoind notifier shutting down...
2024-07-24 22:28:14.057 [ERR] HSWC: InterceptableSwitch stopped: block epoch stream stopped
2024-07-24 22:28:14.057 [ERR] NTFN: dead epoch stream in BestBlockTracker
2024-07-24 22:28:14.057 [ERR] RPCS: [/chainrpc.ChainNotifier/RegisterBlockEpochNtfn]: chain notifier shutting down
2024-07-24 22:28:14.060 [INF] NTFN: Stopping mempool notifier
2024-07-24 22:28:14.060 [INF] CHFT: ChannelEventStore shutting down...
2024-07-24 22:28:14.075 [INF] HLCK: Health monitor shutting down...
2024-07-24 22:28:14.177 [INF] RPCS: Stopping RPC Server
2024-07-24 22:28:14.188 [INF] RPCS: Stopping VersionRPC Sub-RPC Server
2024-07-24 22:28:14.194 [INF] RPCS: Stopping WalletKitRPC Sub-RPC Server
2024-07-24 22:28:14.198 [INF] RPCS: Stopping WatchtowerRPC Sub-RPC Server
2024-07-24 22:28:14.198 [INF] RPCS: Stopping AutopilotRPC Sub-RPC Server
2024-07-24 22:28:14.198 [INF] RPCS: Stopping ChainRPC Sub-RPC Server
2024-07-24 22:28:14.202 [INF] RPCS: Stopping SignRPC Sub-RPC Server
2024-07-24 22:28:14.204 [INF] RPCS: Stopping RouterRPC Sub-RPC Server
2024-07-24 22:28:14.208 [INF] RPCS: Stopping PeersRPC Sub-RPC Server
2024-07-24 22:28:14.208 [INF] RPCS: Stopping WatchtowerClientRPC Sub-RPC Server
2024-07-24 22:28:14.209 [INF] RPCS: Stopping NeutrinoKitRPC Sub-RPC Server
2024-07-24 22:28:14.209 [INF] RPCS: Stopping InvoicesRPC Sub-RPC Server
2024-07-24 22:28:14.209 [INF] TORC: Stopping tor controller
2024-07-24 22:28:14.230 [INF] LTND: Shutdown complete

**Reply by guggero** [reply]
Can you post the full log here? Since the error or action that's causing the shutdown isn't visible.

**Reply by guitarchair** [reply]
[lnd.log](https://github.com/user-attachments/files/16380167/lnd.log)

**Reply by guggero** [reply]
Thanks! This is very weird... It looks like something's shutting down `lnd` right after start. Is it possible there's some script running amok or something wrong with the docker container? I don't think it's a problem with `lnd` itself.

**Reply by guitarchair** [reply]
I only have 4 apps installed: Bitcoin, electrs, lnd, RTL (ride the lightning).
I tried stopping RTL and restart LND but I still get the same stuck loop.
Bitcoind and electrs work fine and 100% sync.
What else could I check that's shutting down lnd? I mean what tool/command could be useful to troubleshoot the source of the problem?

**Reply by guggero** [reply]
Hmm, I don't really have a good idea... Since it kind of stops after trying to open the main database, you could do a manual check if it is healthy. You can do that by stopping `lnd`, then running [`chantools compactdb`](https://github.com/lightninglabs/chantools/blob/master/doc/chantools_compactdb.md) on it (either on Umbrel itself or by copying the `/data/umbrel-os/home/umbrel/umbrel/app-data/lightning/data/lnd/data/graph/mainnet/channel.db` file to another machine and running `chantools` there.

**Reply by guitarchair** [reply]
Since the same problem is happening with both the older database and new fresh database, and in both cases the database works fine right after the installation, but not after a restart, it doesn't seem like a corrupt database file problem...
I'm thinking to perhaps uninstall LND from the web GUI and then install it from cli (the regular lnd version, not the umbrel version). Is that something that might work? I'm just unsure if will cause other problems, like using RTL for opening channels (because RTL is using a web GUI).
I'm also adding the docker inspect file in case that can shed more light. Notice the cmd part in it, maybe these options can give a clue:
"Cmd": [
                "--configfile=/data/.lnd/umbrel-lnd.conf",
                "--listen=0.0.0.0:9735",
                "--rpclisten=0.0.0.0:10009",
                "--restlisten=0.0.0.0:8080",
                "--bitcoin.active",
                "--bitcoin.mainnet",
                "--bitcoin.node=bitcoind",
                "--bitcoind.rpchost=10.21.21.8:8332",
                "--bitcoind.rpcuser=umbrel",
                "--bitcoind.rpcpass=REDUCTED PASSWORD",
                "--bitcoind.zmqpubrawblock=tcp://10.21.21.8:28332",
                "--bitcoind.zmqpubrawtx=tcp://10.21.21.8:28333",
                "--tor.active",
                "--tor.v3",
                "--tor.control=10.21.21.11:29051",
                "--tor.socks=10.21.21.11:9050",
                "--tor.targetipaddress=10.21.21.9",
                "--tor.password=REDUCTED PASSWORD="
            ],

[docker_inspect_lnd.txt](https://github.com/user-attachments/files/16381133/docker_inspect_lnd.txt)

**Reply by guitarchair** [reply, marked as answer]
I found some more interesting logs related to the bbolt database, maybe you can make sense of it?
In the inspect log I found the "log path" which has more info in it:
"LogPath": "/var/lib/docker/containers/8ced8f733de32a96af9b067e814ca13e3053395dc8a8565d69a75bb737a6b07f/8ced8f733de32a96af9b067e814ca13e3053395dc8a8565d69a75bb737a6b07f-json.log"

This is part of the output. Notice the "page" error related to bbolt. So maybe it is a file corruption after all?
I'm going to simply uninstall LND (again), and reinstall and hopefully this time there won't be a db corruption. I'd rather do that then start dealing with compaction of a file that really has nothing on it except for the seed. There are no channels opened yet.

{"log":"2024-07-25 17:10:26.259 [INF] LTND: Active chain: Bitcoin (network=mainnet)\n","stream":"stdout","time":"2024-07-25T17:10:26.259349395Z"}
{"log":"2024-07-25 17:10:26.260 [INF] RPCS: RPC server listening on 0.0.0.0:10009\n","stream":"stdout","time":"2024-07-25T17:10:26.26071092Z"}
{"log":"2024-07-25 17:10:26.275 [INF] RPCS: gRPC proxy started at 0.0.0.0:8080\n","stream":"stdout","time":"2024-07-25T17:10:26.275169707Z"}
{"log":"2024-07-25 17:10:26.275 [INF] LTND: Opening the main database, this might take a few minutes...\n","stream":"stdout","time":"2024-07-25T17:10:26.27546649Z"}
{"log":"2024-07-25 17:10:26.275 [INF] LTND: Opening bbolt database, sync_freelist=false, auto_compact=false\n","stream":"stdout","time":"2024-07-25T17:10:26.275739418Z"}
{"log":"2024-07-25 17:10:26.286 [INF] LTND: Shutdown complete\n","stream":"stdout","time":"2024-07-25T17:10:26.286297227Z"}
{"log":"\n","stream":"stdout","time":"2024-07-25T17:10:26.286313514Z"}
{"log":"panic: assertion failed: Page expected to be: 3436, but self identifies as 9552306040493850340\n","stream":"stderr","time":"2024-07-25T17:10:26.288428026Z"}
{"log":"\n","stream":"stderr","time":"2024-07-25T17:10:26.2884656Z"}
{"log":"goroutine 1 [running]:\n","stream":"stderr","time":"2024-07-25T17:10:26.288474502Z"}
{"log":"go.etcd.io/bbolt._assert(...)\n","stream":"stderr","time":"2024-07-25T17:10:26.288486591Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/db.go:1359\n","stream":"stderr","time":"2024-07-25T17:10:26.288493342Z"}
{"log":"go.etcd.io/bbolt.(*page).fastCheck(0x7f2825b6c000, 0xd6c)\n","stream":"stderr","time":"2024-07-25T17:10:26.288499766Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/page.go:57 +0x1d9\n","stream":"stderr","time":"2024-07-25T17:10:26.288508667Z"}
{"log":"go.etcd.io/bbolt.(*Tx).page(0x334cd28?, 0x1b79dc0?)\n","stream":"stderr","time":"2024-07-25T17:10:26.288515937Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx.go:534 +0x7b\n","stream":"stderr","time":"2024-07-25T17:10:26.288522199Z"}
{"log":"go.etcd.io/bbolt.(*Tx).forEachPageInternal(0xc00017a9a0, {0xc00096c5a0, 0x4, 0xa}, 0xc00079c588)\n","stream":"stderr","time":"2024-07-25T17:10:26.288528569Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx.go:546 +0x5d\n","stream":"stderr","time":"2024-07-25T17:10:26.288535452Z"}
{"log":"go.etcd.io/bbolt.(*Tx).forEachPageInternal(0xc00017a9a0, {0xc00096c5a0, 0x3, 0xa}, 0xc00079c588)\n","stream":"stderr","time":"2024-07-25T17:10:26.288543722Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx.go:555 +0xc8\n","stream":"stderr","time":"2024-07-25T17:10:26.288550472Z"}
{"log":"go.etcd.io/bbolt.(*Tx).forEachPageInternal(0xc00017a9a0, {0xc00096c5a0, 0x2, 0xa}, 0xc00079c588)\n","stream":"stderr","time":"2024-07-25T17:10:26.288557221Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx.go:555 +0xc8\n","stream":"stderr","time":"2024-07-25T17:10:26.288564665Z"}
{"log":"go.etcd.io/bbolt.(*Tx).forEachPageInternal(0xc00017a9a0, {0xc00096c5a0, 0x1, 0xa}, 0xc00079c588)\n","stream":"stderr","time":"2024-07-25T17:10:26.28863353Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx.go:555 +0xc8\n","stream":"stderr","time":"2024-07-25T17:10:26.288645391Z"}
{"log":"go.etcd.io/bbolt.(*Tx).forEachPage(...)\n","stream":"stderr","time":"2024-07-25T17:10:26.288652871Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx.go:542\n","stream":"stderr","time":"2024-07-25T17:10:26.28865892Z"}
{"log":"go.etcd.io/bbolt.(*Tx).checkBucket(0xc00017a9a0, 0xc000934bc0, 0xc00079c990, 0xc00079c8d0, {0x238aea0, 0x350cd40}, 0xc0009984e0)\n","stream":"stderr","time":"2024-07-25T17:10:26.288694638Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx_check.go:83 +0x114\n","stream":"stderr","time":"2024-07-25T17:10:26.288703585Z"}
{"log":"go.etcd.io/bbolt.(*Tx).checkBucket.func2({0x7f2826ffd3bb?, 0xc00079c6c0?, 0xc00044f840?})\n","stream":"stderr","time":"2024-07-25T17:10:26.288710251Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/tx_check.go:110 +0x90\n","stream":"stderr","time":"2024-07-25T17:10:26.2887169Z"}
{"log":"go.etcd.io/bbolt.(*Bucket).ForEachBucket(0x0?, 0xc00079c788)\n","stream":"stderr","time":"2024-07-25T17:10:26.288724261Z"}
{"log":"\u0009go.etcd.io/bbolt@v1.3.7/bucket.go:403 +0x96\n","stream":"stderr","time":"2024-07-25T17:10:26.28873068Z"}
{"log":"go.etcd.io/bbolt.(*Tx).checkBucket(0xc00017a9a0, 0xc00017a9b8, 0xc00079c990, 0xc00079c8d0, {0x238aea0, 0x350cd40}, 0xc0009984e0)\n","stream":"stderr","time":"2024-07-25T17:10:26.288749326Z"}


The full log file here:
[container_log_lnd.txt](https://github.com/user-attachments/files/16381520/container_log_lnd.txt)

**Comment by guggero** [comment]
> So maybe it is a file corruption after all?

Yeah, that's what those errors mean unfortunately. So either it was interrupted unexpectedly (power loss?) or something's not good with the power supply to the hard drive (this sometimes happens with Raspberry Pis, bad power supply issues).

**Reply by guitarchair** [reply]
I'm using a Dell laptop but Umbrel runs entirely on a USB connected SSD (Crucial X9 Pro). There was no power loss.

**Reply by guitarchair** [reply]
Okay I was able to restart the node and now LND is able to start... hopefully this time it won't happen again... or maybe it's better to save a backup of the database and then simply replace a corrupt file.

**Reply by guggero** [reply]
> better to save a backup of the database and then simply replace a corrupt file.

Not if you want to put your channels at risk! Didn't you read the link I sent you above? Channels force closing is just the "optimal" case for running with an old channel state. The worst case is you also lose all funds in those channels. So no, definitely never restore a `channel.db` file from a backup!

**Reply by guitarchair** [reply]
Right, I meant the channel.backup file specifically. Although it's unclear to me, if the file itself is already the backup, why do I need to back it up? And if it is a backup, why couldn't I just use that file instead of uninstalling LND completely and reinstalling it? Or is the "bbolt" database referring to other files?

**Reply by guggero** [reply]
The role of the `channel.backup` file (also known as the SCB file, "Static Channel Backup") is also described in that document.
The idea of backing that file up is to have it in a separate location in case the whole disk is defective.
BUT, the SCB only contains a channel shell, enough information to connect out to your peers and ask them to force close. So you're fully relying on them to close the channel. If they aren't around anymore, you can't get your funds back.
That's why it's so important to make sure the `channel.db` (which is the `bbolt` database) is healthy and doesn't corrupt.

So if you plan on having a non-trivial amount on your node, I'd suggest to have a bit more of a sturdy disk setup than an external, USB connected drive, as that's quite unreliable (just speaking from experience of other node runners having similar difficulties).

**Reply by guitarchair** [reply]
Thanks for the explanation and the help.
If the node fails again I might have to change hardware to internal SSD.
